### ISSUE TITLE
An /v1/embeddings endpoint

### ISSUE THREAD

**Issue by bisand** [bisand/kvad#10]
Serve `/v1/embeddings`: text in, a vector out. Listed as a candidate in [#7](https://github.com/bisand/kvad/issues/7), and much the cheapest thing on that list — with one qualification that decides the whole shape of the work.

## Read this first

`retrieve.rs` already considered embeddings and rejected them, for our own retrieval, with reasons (`crates/llm/src/retrieve.rs:11`): a BERT-class encoder is "a whole architecture, not a feature", and BM25 can say what each term contributed to a match, "which is the thing an embedding cannot do at all".

That reasoning is about *explainable retrieval over a documentation corpus*, and it stands. It says nothing about serving vectors to a client that asked for them, which is a different request with a different answer. Nothing here proposes changing how retrieval works — see the last section.

## The qualification that makes this cheap

There are two families of embedding model, and they cost very different amounts.

**Decoder-based** — Qwen3-Embedding, gte-Qwen2, e5-mistral. These *are* Qwen3 and Llama models. `llama.rs` already claims `llama`, `mistral`, `qwen2` and `qwen3` (`crates/llm/src/model/llama.rs:53`), so the architecture is already in this repository. What makes them embedding models is what you do with the output: pool the hidden states (last token, for these) and L2-normalise. **No new architecture at all.**

**BERT-class** — BGE, GTE, E5, nomic-embed, all-MiniLM. Bidirectional attention, learned position embeddings, token-type embeddings, CLS or mean pooling. This is a new architecture, exactly as `retrieve.rs` says. The bidirectional part is not a detail: `forward_batch` gets its causal mask *for free* because query `i` attends over `pos0 + i + 1` positions, and an encoder that must see its own future breaks that property.

**So: start with a decoder-based model.** Qwen3-Embedding-0.6B is small, current and already an architecture we run. BERT-class is a separate, later issue if somebody wants those checkpoints.

## The one engine change, and it is small

`Session::forward` returns logits. An embedding needs the hidden state *before* the output head.

That seam already exists in every architecture. `run_batch` runs every block and returns the residual stream for all `m` positions with no output head applied (`crates/llm/src/model/llama.rs:164`, `gpt2.rs:146`, `deepseek.rs:679`) — it was split out because generation wants the last position's logits and scoring wants all of them. Then:

```rust
// forward_batch_all, llama.rs:295
let mut xs = self.run_batch(tokens, cache);
for i in 0..m {
    let normed = rms_norm(&xs[i * e..(i + 1) * e], &self.final_norm, self.spec.eps);
    xs[i * e..(i + 1) * e].copy_from_slice(&normed);
}
self.lm_head.as_ref().unwrap_or(&self.embed).matmul_bt(&xs, m, None)
```

What an embedding wants is that function **without the last line**. A `hidden_batch` on `Transformer`, mirrored as a `hidden` on `Session`, is a handful of lines per architecture, and `forward_batch_all` is its own test oracle — the same pattern the existing batched methods already use.

It is also cheaper than the alternative, not just tidier: the head is `vocab_size × n_embd`, the largest single matmul in the model, and this skips it entirely.

## A correctness trap

`Llm` reuses the KV cache across calls by common prefix (`crates/llm/src/runtime.rs:312`). Embedding inputs are independent — the second text is not a continuation of the first — so a cache carried between them would silently produce wrong vectors, which is the worst failure mode available here because nothing looks broken.

Each input needs a clean cache (`Llm::reset`, `runtime.rs:254`), and the test should be that embedding the same text first and tenth in a batch gives bit-identical vectors.

## The endpoint

OpenAI's shape, same discipline as `/v1/chat/completions`:

- `input` accepts a string, an array of strings, or arrays of token ids.
- `encoding_format`: `float` and `base64`.
- `dimensions`: Qwen3-Embedding supports Matryoshka truncation, which means truncate then **re-normalise** — getting the renormalise wrong gives vectors that are subtly wrong rather than obviously wrong.
- `usage` with `prompt_tokens` and `total_tokens`.
- Inputs longer than the context window: OpenAI errors, most servers truncate. Pick one and say so, rather than inheriting whatever `n_ctx` assertion fires first.

Two things that are easy to get wrong and silent when wrong:

- **Pooling and normalisation must match the model's own convention.** Last-token for the decoder family, and L2 at the end.
- **Query and document asymmetry.** Qwen3-Embedding expects an instruction prefix on queries and not on documents. A server that applies the wrong one degrades retrieval quality with no error anywhere.

Both want verifying as **cosine similarity against reference vectors** from the reference implementation, on fixed inputs, not by whether the nearest neighbours look plausible.

## Batching

Embedding a hundred texts is a hundred independent prefills, and prefill is already the batched path — `forward_batch` exists precisely because feeding tokens together reads each weight once instead of once per token. Within one input that win is already ours. Across inputs it is not: with a cache reset between them, a hundred texts are a hundred sequential passes today.

That makes this a natural consumer of [#5](https://github.com/bisand/kvad/issues/5) and a much simpler one than chat — no sampling, no termination conditions, no continuous admission, every sequence the same length of work and all of them finishing together. It may be worth doing embeddings batching first for exactly that reason: it is the same idea with none of the scheduling.

## Where it touches the rest

- **[#2](https://github.com/bisand/kvad/issues/2)** — an embedding model is small and is precisely the thing you want resident *alongside* a chat model rather than instead of it. This is a good argument for that issue and a poor experience without it.
- **[#7](https://github.com/bisand/kvad/issues/7)** — `/v1/models` has to say which models can be embedded with, the way it already reports `tools`.
- **[#3](https://github.com/bisand/kvad/issues/3)** — `kvad embed`.

## Retrieval, explicitly left alone

Once an encoder exists, hybrid retrieval — BM25 for the terms that are strings, dense for the questions that are paraphrases — becomes testable rather than theoretical. That is worth measuring against the current index on a real corpus some day.

It is not this issue, and BM25 should keep its place regardless: `Index::search` saying what each term contributed is a property worth more than a few points of recall on a documentation corpus, and it is the reason the current design is what it is.

## Suggested order

1. `hidden_batch` on `Transformer` and `hidden` on `Session`, checked against `forward_batch_all` minus the head.
2. Pooling, L2 normalisation and the cache reset, with the first-and-tenth identity test.
3. `/v1/embeddings` against Qwen3-Embedding-0.6B, verified by cosine similarity to reference vectors.
4. `dimensions`, `base64`, the truncation rule, and the query/document prefixes.
5. Batching across inputs.
6. BERT-class encoders, as their own issue, if the checkpoints are wanted.
